### ISSUE TITLE
Invalid value for DatastoreMappingParams.target_profile when deploying template from Content Library

### ISSUE THREAD

**Issue by BrandonStiff** [hashicorp/terraform-provider-vsphere#1525]
### Terraform Version

0.15.3

### vSphere Provider Version

2.0.2

### vSphere Version

vSphere 6.7.0 Build 14368073

### Affected Resource(s)

- `vsphere_virtual_machine`

### Terraform Configuration Files

```hcl
#  main.tf
terraform {
  required_version = "~> 0.15.3"

  required_providers {
    vsphere = {
      source  = "hashicorp/vsphere"
      version = "~> 2.0"
    }
  }
}

variable "name" {
  type = string
}

variable "vapp_container" {
  type = string
}

variable "datacenter" {
  type = string
}

variable "storage_policy_name" {
  type = string
}

variable "network_name" {
  type = string
}

variable "disk_size_gb" {
  type = number
}

variable "vsphere_template" {
  type = string
}

variable "content_library" {
  type = string
}

variable "vsphere_server" {
  type = string
}

provider "vsphere" {
  vsphere_server       = var.vsphere_server
  allow_unverified_ssl = true
}

data "vsphere_datacenter" "dc" {
  name = var.datacenter
}

data "vsphere_storage_policy" "storage_policy" {
  name = var.storage_policy_name
}

data "vsphere_network" "network" {
  name                            = var.network_name
  datacenter_id                   = data.vsphere_datacenter.dc.id
  distributed_virtual_switch_uuid = ""
}

data "vsphere_content_library" "library" {
  name = var.content_library
}

data "vsphere_content_library_item" "template" {
  name       = var.vsphere_template
  library_id = data.vsphere_content_library.library.id
  type       = "ovf"
}

data "vsphere_vapp_container" "vapp_container" {
  name          = var.vapp_container
  datacenter_id = data.vsphere_datacenter.dc.id
}

resource "vsphere_virtual_machine" "vm" {
  name              = var.name
  resource_pool_id  = data.vsphere_vapp_container.vapp_container.id
  storage_policy_id = data.vsphere_storage_policy.storage_policy.id
  guest_id          = "ubuntu64Guest"
  scsi_type         = "lsilogic"
  annotation        = "Created by Terraform - ${timestamp()}"
  num_cpus          = 4
  memory            = 14000
  enable_disk_uuid  = true

  network_interface {
    network_id   = data.vsphere_network.network.id
    adapter_type = "vmxnet3"
  }

  disk {
    label            = "disk0"
    size             = var.disk_size_gb
    eagerly_scrub    = false
    thin_provisioned = true
  }

  clone {
    template_uuid = data.vsphere_content_library_item.template.id
    timeout       = 30
  }
}
```

Then supply a terraform.tfvars` that looks like this:

```hcl
# terraform.tfvars
vsphere_server      = "vcenter-server.local"
name                = "vm-name"
vapp_container      = "vapp-container-name"
storage_policy_name = "storage-policy-name"
network_name        = "vnet-name"
datacenter          = "vsphere-datacenter"
disk_size_gb        = 108
vsphere_template    = "content-library-template-name"
content_library     = "content-library-name"
```

### Debug Output

[Debug Output](https://gist.github.com/BrandonStiff/ed8d10a3addff4b720f9457ff81fb7c4)

### Expected Behavior

It should have stood up the VM in the vApp using the storage datastores indicated by the storage policy specified in the variables.

### Actual Behavior

It fails with the following output:

```text
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vsphere_virtual_machine.vm: Creating...
vsphere_virtual_machine.vm: Still creating... [10s elapsed]
vsphere_virtual_machine.vm: Still creating... [20s elapsed]
vsphere_virtual_machine.vm: Still creating... [30s elapsed]
vsphere_virtual_machine.vm: Still creating... [40s elapsed]
vsphere_virtual_machine.vm: Still creating... [50s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m0s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m10s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m20s elapsed]
vsphere_virtual_machine.vm: Still creating... [1m30s elapsed]
╷
│ Error: deploy error: Invalid value for DatastoreMappingParams.target_profile: e44aa044-5d3b-4793-865a-d9947d8c9234.
│
│   with vsphere_virtual_machine.vm,
│   on main.tf line 85, in resource "vsphere_virtual_machine" "vm":
│   85: resource "vsphere_virtual_machine" "vm" {
│
```

### Steps to Reproduce

1. Create a [tag-based vSphere storage policy](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.storage.doc/GUID-D025AA68-BF00-4FC2-9C7E-863E5787E743.html) that encompasses at least one datastore.
2. Create a VM template in the vcenter [content library](https://docs.vmware.com/en/VMware-vSphere/7.0/com.vmware.vsphere.vm_admin.doc/GUID-254B2CE8-20A8-43F0-90E8-3F6776C2C896.html) that can be used as a clone base.
3. Create a new vApp.
4. Create `maint.f` with the contents above.
5. Create `terraform.tfvars` and paste the contents I indicated above.
6. Adjust the variable values to match your test environment.
7. Set VSPHERE_USERNAME and VSPHERE_PASSWORD environment variables to your vcenter server credentials.
8. `terraform init`
9. `terraform apply`

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

**Comment by tenthirtyam** [comment]
@BrandonStiff -

I've tested this in my environment w/o issue using VMware vSphere 7.0 Update 2 and the following without issue.

In the test scenario, I'm deploying two VMs (`foo` and `bar`) to a vApp (`hello-world`) with each VM being targeted for a storage policy based on a a tag that corresponds to the the VM names (`foo` and `bar`).

```console
Terraform v1.0.11
on linux_amd64
+ provider registry.terraform.io/hashicorp/vsphere v2.0.2
```

```console
Storage Policies based on Tags:
- foo => local-ssd-01
- bar => local-ssd-02
```

**`versions.tf`**

```hcl
terraform {
  required_providers {
    vsphere = {
      source  = "hashicorp/vsphere"
      version = ">= 2.0.2"
    }
  }
  required_version = ">= 1.0.11"
}
```

**`variables.tf`**

```hcl
# Credentials

variable "vsphere_server" {
  type = string
}

variable "vsphere_username" {
  type = string
}

variable "vsphere_password" {
  type = string
}

variable "vsphere_insecure" {
  type    = bool
  default = false
}

# vSphere Settings

variable "vsphere_datacenter" {
  type = string
}

variable "vsphere_cluster" {
  type = string
}

variable "vsphere_folder" {
  type = string
}

variable "vsphere_network" {
  type = string
}

variable "vsphere_content_library" {
  type = string
}

variable "vsphere_content_library_ovf" {
  type = string
}

variable "vsphere_vapp_container" {
  type = string
}

# Virtual Machine Settings

variable "vm_cpus" {
  type = number
}

variable "vm_memory" {
  type = number
}

variable "vm_disk_size" {
  type = number
}

variable "vm_firmware" {
  type = string
}

variable "vm_efi_secure_boot_enabled" {
  type = bool
}

variable "vm_names" {
  type = map(any)
}
```

**`main.tf`**

```hcl
provider "vsphere" {
  vsphere_server       = var.vsphere_server
  user                 = var.vsphere_username
  password             = var.vsphere_password
  allow_unverified_ssl = var.vsphere_insecure
}

data "vsphere_datacenter" "datacenter" {
  name = var.vsphere_datacenter
}

data "vsphere_network" "network" {
  name          = var.vsphere_network
  datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_compute_cluster" "cluster" {
  name          = var.vsphere_cluster
  datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_storage_policy" "storage_policy" {
  for_each      = var.vm_names
  name          = each.value["name"]
}

data "vsphere_content_library" "content_library" {
  name = var.vsphere_content_library
}

data "vsphere_content_library_item" "content_library_item" {
  name       = var.vsphere_content_library_ovf
  type       = "ovf"
  library_id = data.vsphere_content_library.content_library.id
}

data "vsphere_vapp_container" "vapp_container" {
  name          = var.vsphere_vapp_container
  datacenter_id = data.vsphere_datacenter.datacenter.id
}

resource "vsphere_virtual_machine" "vm" {
  for_each                = var.vm_names
  name                    = each.value["name"]
  folder                  = var.vsphere_folder
  num_cpus                = var.vm_cpus
  memory                  = var.vm_memory
  firmware                = var.vm_firmware
  efi_secure_boot_enabled = var.vm_efi_secure_boot_enabled
  storage_policy_id       = data.vsphere_storage_policy.storage_policy[each.key].id
  resource_pool_id        = data.vsphere_vapp_container.vapp_container.id
  network_interface {
    network_id = data.vsphere_network.network.id
  }
  disk {
    label            = "disk0"
    size             = var.vm_disk_size
    thin_provisioned = true
  }
  clone {
    template_uuid = data.vsphere_content_library_item.content_library_item.id
    customize {
      linux_options {
        host_name = each.value["name"]
        domain    = each.value["domain"]
      }
      network_interface {
        ipv4_address = each.value["ipv4_address"]
        ipv4_netmask = each.value["ipv4_netmask"]
      }
      ipv4_gateway    = each.value["ipv4_gateway"]
      dns_suffix_list = each.value["dns_suffix_list"]
      dns_server_list = each.value["dns_server_list"]
    }
  }
  lifecycle {
    ignore_changes = [
      clone[0].template_uuid,
    ]
  }
}
```

**`terraform.tfvars`**

```hcl
# Credentials

vsphere_server   = "m01-vc01.rainpole.io"
vsphere_username = "administrator@vsphere.local"
vsphere_password = "VMware1!"
vsphere_insecure = true

# vSphere Settings

vsphere_datacenter          = "m01-dc01"
vsphere_cluster             = "m01-cl01"
vsphere_folder              = ""
vsphere_network             = "M - 172.16.11.0"
vsphere_content_library     = "m01-lib01"
vsphere_content_library_ovf = "linux-ubuntu-server-20-04-lts"
vsphere_vapp_container      = "hello-world"

# Virtual Machines Settings

vm_cpus                    = 2
vm_memory                  = 4096
vm_disk_size               = 60
vm_firmware                = "efi"
vm_efi_secure_boot_enabled = true
vm_names = {
  foo_vm = {
    name            = "foo",
    domain          = "rainpole.io"
    ipv4_address    = "172.16.11.101"
    ipv4_netmask    = 24
    ipv4_gateway    = "172.16.11.1"
    dns_suffix_list = ["rainpole.io"]
    dns_server_list = ["172.16.11.11", "172.16.11.12"]

  },
  bar_vm = {
    name            = "bar",
    domain          = "rainpole.io"
    ipv4_address    = "172.16.11.102"
    ipv4_netmask    = 24
    ipv4_gateway    = "172.16.11.1"
    dns_suffix_list = ["rainpole.io"]
    dns_server_list = ["172.16.11.11", "172.16.11.12"]
  }
}

```
Results:

```console
$ terraform apply --auto-approve

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # vsphere_virtual_machine.vm["bar_vm"] will be created
  + resource "vsphere_virtual_machine" "vm" {
      + boot_retry_delay                        = 10000
      + change_version                          = (known after apply)
      + cpu_limit                               = -1
      + cpu_share_count                         = (known after apply)
      + cpu_share_level                         = "normal"
      + datastore_id                            = (known after apply)
      + default_ip_address                      = (known after apply)
      + efi_secure_boot_enabled                 = true
      + ept_rvi_mode                            = "automatic"
      + firmware                                = "efi"
      + force_power_off                         = true
      + guest_id                                = (known after apply)
      + guest_ip_addresses                      = (known after apply)
      + hardware_version                        = (known after apply)
      + host_system_id                          = (known after apply)
      + hv_mode                                 = "hvAuto"
      + id                                      = (known after apply)
      + ide_controller_count                    = 2
      + imported                                = (known after apply)
      + latency_sensitivity                     = "normal"
      + memory                                  = 4096
      + memory_limit                            = -1
      + memory_share_count                      = (known after apply)
      + memory_share_level                      = "normal"
      + migrate_wait_timeout                    = 30
      + moid                                    = (known after apply)
      + name                                    = "bar"
      + num_cores_per_socket                    = 1
      + num_cpus                                = 2
      + poweron_timeout                         = 300
      + reboot_required                         = (known after apply)
      + resource_pool_id                        = "resgroup-v25022"
      + run_tools_scripts_after_power_on        = true
      + run_tools_scripts_after_resume          = true
      + run_tools_scripts_before_guest_shutdown = true
      + run_tools_scripts_before_guest_standby  = true
      + sata_controller_count                   = 0
      + scsi_bus_sharing                        = "noSharing"
      + scsi_controller_count                   = 1
      + scsi_type                               = "pvscsi"
      + shutdown_wait_timeout                   = 3
      + storage_policy_id                       = "76e05bb9-66f8-4698-9bfd-ad9a2cfd10df"
      + swap_placement_policy                   = "inherit"
      + uuid                                    = (known after apply)
      + vapp_transport                          = (known after apply)
      + vmware_tools_status                     = (known after apply)
      + vmx_path                                = (known after apply)
      + wait_for_guest_ip_timeout               = 0
      + wait_for_guest_net_routable             = true
      + wait_for_guest_net_timeout              = 5

      + clone {
          + template_uuid = "93d636ab-74c7-45d8-8795-e00997fa9755"
          + timeout       = 30

          + customize {
              + dns_server_list = [
                  + "172.16.11.11",
                  + "172.16.11.12",
                ]
              + dns_suffix_list = [
                  + "rainpole.io",
                ]
              + ipv4_gateway    = "172.16.11.1"
              + timeout         = 10

              + linux_options {
                  + domain       = "rainpole.io"
                  + host_name    = "bar"
                  + hw_clock_utc = true
                }

              + network_interface {
                  + ipv4_address = "172.16.11.102"
                  + ipv4_netmask = 24
                }
            }
        }

      + disk {
          + attach            = false
          + controller_type   = "scsi"
          + datastore_id      = "<computed>"
          + device_address    = (known after apply)
          + disk_mode         = "persistent"
          + disk_sharing      = "sharingNone"
          + eagerly_scrub     = false
          + io_limit          = -1
          + io_reservation    = 0
          + io_share_count    = 0
          + io_share_level    = "normal"
          + keep_on_remove    = false
          + key               = 0
          + label             = "disk0"
          + path              = (known after apply)
          + size              = 60
          + storage_policy_id = (known after apply)
          + thin_provisioned  = true
          + unit_number       = 0
          + uuid              = (known after apply)
          + write_through     = false
        }

      + network_interface {
          + adapter_type          = "vmxnet3"
          + bandwidth_limit       = -1
          + bandwidth_reservation = 0
          + bandwidth_share_count = (known after apply)
          + bandwidth_share_level = "normal"
          + device_address        = (known after apply)
          + key                   = (known after apply)
          + mac_address           = (known after apply)
          + network_id            = "network-1001"
        }
    }

  # vsphere_virtual_machine.vm["foo_vm"] will be created
  + resource "vsphere_virtual_machine" "vm" {
      + boot_retry_delay                        = 10000
      + change_version                          = (known after apply)
      + cpu_limit                               = -1
      + cpu_share_count                         = (known after apply)
      + cpu_share_level                         = "normal"
      + datastore_id                            = (known after apply)
      + default_ip_address                      = (known after apply)
      + efi_secure_boot_enabled                 = true
      + ept_rvi_mode                            = "automatic"
      + firmware                                = "efi"
      + force_power_off                         = true
      + guest_id                                = (known after apply)
      + guest_ip_addresses                      = (known after apply)
      + hardware_version                        = (known after apply)
      + host_system_id                          = (known after apply)
      + hv_mode                                 = "hvAuto"
      + id                                      = (known after apply)
      + ide_controller_count                    = 2
      + imported                                = (known after apply)
      + latency_sensitivity                     = "normal"
      + memory                                  = 4096
      + memory_limit                            = -1
      + memory_share_count                      = (known after apply)
      + memory_share_level                      = "normal"
      + migrate_wait_timeout                    = 30
      + moid                                    = (known after apply)
      + name                                    = "foo"
      + num_cores_per_socket                    = 1
      + num_cpus                                = 2
      + poweron_timeout                         = 300
      + reboot_required                         = (known after apply)
      + resource_pool_id                        = "resgroup-v25022"
      + run_tools_scripts_after_power_on        = true
      + run_tools_scripts_after_resume          = true
      + run_tools_scripts_before_guest_shutdown = true
      + run_tools_scripts_before_guest_standby  = true
      + sata_controller_count                   = 0
      + scsi_bus_sharing                        = "noSharing"
      + scsi_controller_count                   = 1
      + scsi_type                               = "pvscsi"
      + shutdown_wait_timeout                   = 3
      + storage_policy_id                       = "a374b0e1-940c-497b-9dec-3ff14baff199"
      + swap_placement_policy                   = "inherit"
      + uuid                                    = (known after apply)
      + vapp_transport                          = (known after apply)
      + vmware_tools_status                     = (known after apply)
      + vmx_path                                = (known after apply)
      + wait_for_guest_ip_timeout               = 0
      + wait_for_guest_net_routable             = true
      + wait_for_guest_net_timeout              = 5

      + clone {
          + template_uuid = "93d636ab-74c7-45d8-8795-e00997fa9755"
          + timeout       = 30

          + customize {
              + dns_server_list = [
                  + "172.16.11.11",
                  + "172.16.11.12",
                ]
              + dns_suffix_list = [
                  + "rainpole.io",
                ]
              + ipv4_gateway    = "172.16.11.1"
              + timeout         = 10

              + linux_options {
                  + domain       = "rainpole.io"
                  + host_name    = "foo"
                  + hw_clock_utc = true
                }

              + network_interface {
                  + ipv4_address = "172.16.11.101"
                  + ipv4_netmask = 24
                }
            }
        }

      + disk {
          + attach            = false
          + controller_type   = "scsi"
          + datastore_id      = "<computed>"
          + device_address    = (known after apply)
          + disk_mode         = "persistent"
          + disk_sharing      = "sharingNone"
          + eagerly_scrub     = false
          + io_limit          = -1
          + io_reservation    = 0
          + io_share_count    = 0
          + io_share_level    = "normal"
          + keep_on_remove    = false
          + key               = 0
          + label             = "disk0"
          + path              = (known after apply)
          + size              = 60
          + storage_policy_id = (known after apply)
          + thin_provisioned  = true
          + unit_number       = 0
          + uuid              = (known after apply)
          + write_through     = false
        }

      + network_interface {
          + adapter_type          = "vmxnet3"
          + bandwidth_limit       = -1
          + bandwidth_reservation = 0
          + bandwidth_share_count = (known after apply)
          + bandwidth_share_level = "normal"
          + device_address        = (known after apply)
          + key                   = (known after apply)
          + mac_address           = (known after apply)
          + network_id            = "network-1001"
        }
    }

Plan: 2 to add, 0 to change, 0 to destroy.
vsphere_virtual_machine.vm["foo_vm"]: Creating...
vsphere_virtual_machine.vm["bar_vm"]: Creating...
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [10s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [10s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [20s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [20s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [30s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [30s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [40s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [40s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [50s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [50s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [1m0s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [1m0s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [1m10s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [1m10s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [1m20s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [1m20s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [1m30s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [1m30s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [1m40s elapsed]
vsphere_virtual_machine.vm["bar_vm"]: Still creating... [1m40s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Still creating... [1m50s elapsed]
vsphere_virtual_machine.vm["foo_vm"]: Creation complete after 1m50s [id=42022dea-6191-dd00-43e5-a51f63651920]
vsphere_virtual_machine.vm["bar_vm"]: Creation complete after 1m50s [id=42024e49-058b-27ef-e7df-da1acbb3cefc]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

UI Screenshots:

<img width="860" alt="image" src="https://user-images.githubusercontent.com/7771363/144138022-ef5058c5-2def-446c-82fc-90f1b1c620d0.png">

<img width="859" alt="image" src="https://user-images.githubusercontent.com/7771363/144138033-f981d184-45c0-4744-ab54-ff21b2519f35.png">

<img width="860" alt="image" src="https://user-images.githubusercontent.com/7771363/144138042-4354cacc-2677-4a58-97d9-181e1c1365ae.png">

<img width="858" alt="image" src="https://user-images.githubusercontent.com/7771363/144138048-6b18a723-a448-490c-8665-3294867d04ec.png">

<img width="858" alt="image" src="https://user-images.githubusercontent.com/7771363/144138066-e1dd8ee7-db4c-4051-831a-960f1706e69b.png">

Hope this helps,
Ryan

**Comment by BrandonStiff** [comment]
Thanks for the response.  I wish I had a VMware 7.0 environment to run this test against.  I'll see if I can get my hands on one.  That's the only major difference I see in our tests.

**Comment by tenthirtyam** [comment]
Happy to help, Brandon. Let me know if the example above works on your 6.7 - hope it will help you out.

**ProTip**: if you don’t have an available testbed, quite often can always use use one of our (me `=` VMware) Hands-on Labs to spin up a pre-built test environment online, install the Terraform binary, and give it go. 

Ryan

**Comment by BrandonStiff** [comment]
Thanks!  I'll give your example a go and report back.

**Comment by BrandonStiff** [comment]
Ok, your example worked.  In digging through all of this, I think I found the issue.  I noticed that the specific cluster I targeted did not have any datastores with the tag specified by the storage policy (other clusters do, but not this specific one).  I somehow missed this in all my digging, and the error did not immediately put two and two together for me.  I'm going to correct this and see if it fixes the issue.
